.vscode/launch.json

```diff
@@ -0,0 +1,31 @@
+{
+    // Use IntelliSense to learn about possible attributes.
+    // Hover to view descriptions of existing attributes.
+    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
+    "version": "0.2.0",
+    "configurations": [
+        {
+            // https://code.visualstudio.com/docs/cpp/config-msvc
+            // https://learn.microsoft.com/en-us/windows-hardware/drivers/debugger/debugger-download-tools
+            "name": "(windbg) Launch",
+            "type": "cppvsdbg",
```

Comment:
similar, this one is sadly not free; you may want to use clangd for intellisense and native-debug for debugging when you want to add those to the repo

Reply:
This is the launcher for debugger, it has nothing to do with `clangd` as it is the language server for code navigation, and I'll leave `clangd` out since its interfering with IntelliSense, may cause problems for others too, IntelliSense will be removed from the workspace recommended `extensions.json`.

The debugger may be problematic, I'm not entirely sure about the `free` licensing you're referring to or what `native-debug` is, however this is the only debugger for MSVC which is what we're using, its not entirely `WinDbg` per se, but the same free proprietary tools Windows SDK bundles - Debugging TOols for Windows, WinDbg, CDB, KD, NTSD...

On a side note, did you mean open source? Please clarify

Reply:
@GitMensch ,
is your "free" referring to free-as-in-speech (libre)? [WinDbg](https://learn.microsoft.com/en-us/windows-hardware/drivers/debugger/debugger-download-tools) is part of Windows SDK which is free-as-in-beer (gratis), but its the only debugger for MSVC, its just configuration and we are not even redistributing this to users or developers

Reply:
This is about the extension necessary - the cppvsdbg debug type is only part of a Microsoft extension that can only be used with Visual Studio Code, so "free beer if you sit only on a specific chair with no single adjustment, that comes pre-configured to tell the maker whenever you sit on it" (and the beer likely doing the same); and with the adjusted recommendations we don't have that available any more, I think. Am I wrong?

Note: If I remember correctly you can use LLDB to debug msvc generated modules, too - for this setup there are some free extensions (like "free to use for anything, anywhere, including adjusting the extension and re-distribute it"- "libre"), too.

Reply:
Nope, its a known issue which is why its never published in the vscode documentation, it may be possible in the CLI/TUI but LLDB cannot be used to debug MSVC-generated symbols in vscode as it [lacks LLDB-MI](https://github.com/microsoft/vscode-cpptools/issues/6617) for non-Apple platforms at the moment, the closest  LLDB for vscode is the [CodeLLDB](https://github.com/vadimcn/codelldb) extension which is not compatible with MSVC-based symbols; not even [MSVC-based Clang](https://github.com/llvm/llvm-project/releases/latest) programs, MSVC-based Clang programs will still need to use WinDbg since its PDB symbols are MSVC-based...
Getting [LLDB-MI](https://github.com/lldb-tools/lldb-mi) to work on Windows will require quite some effort from the community, which is inactive at the moment...

Yes, the recommended beer will be gone from the recommended extensions, I think having this debug configuration is not as evil as telling devs "the free beer is required or recommended", for now the codebase is only compatible with this beer (MSVC), if you need to debug, this is the only available beer (which we did not bundle) at the moment :speak_no_evil:

P.S. I can try to get the codebase compatible with [WinLibs](https://winlibs.com/)/UCRT after this, just not within this PR as it requires quite some changes

Reply:
Note: Just tested, debugging x64 cl.exe generated programs (pdb) with the codelldb extension works just fine - as well as debugging msys2 mingw64 generated executables.

So, if you want to add a debug configuration what do you think about configuring that and add it to the suggested extensions?

Looking on the internet msys2 mingw-w64-x86_64-lldb-mi (if you don't have llvm already setup then the dependencies lead to a 916MB install!) seems to work fine for both GNU and PDB, too - using any vscode extension that supports this, like Native Debug.

Reply:
Can you check out strings and vectors?

Reply:
C strings did work, as I'm currently working on C workspaces only I have no setup for `std::` - but if you get down to test them please drop a note.

Reply:
C strings will definitely work, STLs don't, since upstream are GNU guys and they don't like C++ or STLs, I can accept codeLLDB without being able to debug `std::string` and `std::vector` in this workspace, I'll default this to codeLLDB

Reply:
LLDB folks are GNU guys?
GNU guys don't like C++or STLs (just have a look what some of the GNU software is written in...)?

Note: for GDB you commonly want pretty-print handlers for C++ and STL types, maybe this is similar with LLDB?

Reply:
> The debugger may be problematic, I'm not entirely sure about the `free` licensing you're referring to

the [microsoft extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode.cpptools) is not free, as you can [see in its license](https://marketplace.visualstudio.com/items/ms-vscode.cpptools/license)

> You may install and use any number of copies of the software only with Microsoft [...] Visual Studio Code

--> this explicit disallows use with any other built of vscode repo and the extension also checks for the Microsoft identifiers in product.json, so disables itself when it "believes" it is not run on the "single blessed binary product"... and they even have a wiki page explaining that: https://github.com/microsoft/vscode-cpptools/wiki/Microsoft-Native-Debugger-licensing-and-Microsoft-Visual-Studio-Code

> or what `native-debug` is

It is one of the extensions that can work with lldb-mi, so _if_ you have a binary lldb-mi, then you can use it.

> however this is the only debugger for MSVC which is what we're using,

No, as noted you can use the lldb extension which distributes windows binaries, too (but not lldb-mi) to debug MSVC generated binaries as well as any debug extension that supports lldb-mi.

My point was _exactly_ that I don't this repo to suggest people to use the non-free extension (that will only work on "MS' blessed binaries, so for example not any of my vscode installations)and let all others have a note in their problems pane that the debug type is unsupported.

Reply:
Thank you so much for going above and beyond to clarify, now I understood what you meant, it is free if user/dev is on MS tool suites, not free for any other variants or forks of `Code - OSS`
TL;DR  the rule for this free beer is MS binaries

Reply:
> TL;DR the rule for this free beer is MS binaries

yes, along with binaries where the code is explicit not released - and both have an explicit opt-out of telemetry (which you can set, but as the sources are closed you can only check what goes out by firewall [blocking some of that works, others broke the extension, but that's a too unrelated thing to this PR and I also have no recent data about that, so "ignore that part of my rant, please - just keep in mind 'nobody but MS can actually check what the software does'"]).